File: csharp/838.push-dominoes.cs
/*
 * @lc app=leetcode id=838 lang=csharp
 *
 * [838] Push Dominoes
 */

using System.Collections.Generic;
using System.Text;
using System.Linq;

// @lc code=start
public class Solution {
    public string PushDominoes(string dominoes) {
        var groups = new List<(char State, int Count)>();
        var sb = new StringBuilder(dominoes.Length);
        char prev = '*';
        int len = 1;
        foreach (var c in Enumerable.Concat(dominoes, "*$")) {
            if (c != prev) {
                groups.Add((State: prev, Count: len));
                len = 1;
            } else {
                ++len;
            }
            prev = c;
        }
        int n = groups.Count;
        for (int i = 1; i < n - 1; ++i) {
            (char state, int count) = groups[i];
            if (state == 'L' || state == 'R') {
                sb.Append(state, count);
                continue;
            }
            bool lr = groups[i - 1].State == 'R', rl = groups[i + 1].State == 'L';
            if (lr) {
                if (rl) {
                    int h = count / 2;
                    sb.Append('R', h);
                    if (count % 2 > 0) {
                        sb.Append('.');
                    }
                    sb.Append('L', h);
                } else {
                    sb.Append('R', count);
                }
            } else {
                sb.Append(rl ? 'L' : '.', count);
            }
        }
        return sb.ToString();
    }
}
// @lc code=end

File: java/838.push-dominoes.java
/*
 * @lc app=leetcode id=838 lang=java
 *
 * [838] Push Dominoes
 */

// @lc code=start
class Solution {
    public String pushDominoes(String dominoes) {
        int n = dominoes.length();
        char[] array = dominoes.toCharArray();
        for (int i = 0; i < n;) {
            int ni = i + 1;
            while (ni < n && array[i] == array[ni]) {
                ++ni;
            }
            if (array[i] == '.') {
                boolean r = i > 0 && array[i - 1] == 'R', l = ni < n && array[ni] == 'L';
                if (l) {
                    if (r) {
                        int h = (ni - i) / 2;
                        for (int j = 0; j < h; ++j) {
                            array[i + j] = 'R';
                            array[ni - 1 - j] = 'L';
                        }
                    } else {
                        for (int j = i; j < ni; ++j) {
                            array[j] = 'L';
                        }
                    }
                } else {
                    if (r) {
                        for (int j = i; j < ni; ++j) {
                            array[j] = 'R';
                        }
                    }
                }
            }
            i = ni;
        }
        return new String(array);
    }
}
// @lc code=end

